src/openlifu/plan/protocol.py

```diff
@@ -80,17 +101,198 @@ def to_json(self, compact:bool) -> str:
+            simulate: bool
+                Enable solution simulation (Default: true).
+            scale: bool
+                Triggers solution and simulation scaling to the requested pressure (Default: true).
+            sim_options : sim.SimSetup
+                The options for the k-wave simulation (Default: self.sim_setup).
+            analysis_options: plan.solution.SolutionAnalysisOptions
+                The options for the solution analysis (Default: self.analysis_options).
+            on_pulse_mismatch: plan.protocol.OnPulseMismatchAction
+                An action to take if the number of pulses in the sequence does not match
+                the number of foci (Default: OnPulseMismatchAction.ERROR).
+
+        Returns: Tuple[Solution, xa.DataArray]
+            Represents the solution object including its analysis, and aggregated simulation output.
+        """
+        if sim_options is None:
+            sim_options = self.sim_setup
+        if analysis_options is None:
+            analysis_options = self.analysis_options
+        # check before if target is within bounds
+        self.check_target(target)
+        params, transducer, target = sim_options.setup_sim_scene(transducer, transform, target, self.seg_method, volume=volume, units="m")
+
+        delays_to_stack: List[np.ndarray] = []
+        apodizations_to_stack: List[np.ndarray] = []
+        simulation_outputs_to_stack: List[xa.Dataset] = []
+        simulation_output_stacked: xa.Dataset = None
+        simulation_result_aggregated: xa.Dataset = None
+        foci: List[Point] = self.focal_pattern.get_targets(target)
+        simulation_cycles = np.max([np.round(self.pulse.duration * self.pulse.frequency), 20])
+        out_solution: Solution = Solution()
+
+        # updating solution sequence if pulse mismatch
+        solution_sequence = deepcopy(self.sequence)
+        if (self.sequence.pulse_count % len(foci)) != 0:
+            if on_pulse_mismatch is OnPulseMismatchAction.ERROR:
+                raise ValueError(f"Pulse Count {self.sequence.pulse_count} is not a multiple of the number of foci {len(foci)}")
+            else:
+                if on_pulse_mismatch is OnPulseMismatchAction.ROUND:
+                    solution_sequence.pulse_count = round(self.sequence.pulse_count / len(self.foci)) * len(foci)
+                elif on_pulse_mismatch is OnPulseMismatchAction.ROUNDUP:
+                    solution_sequence.pulse_count = math.ceil(self.sequence.pulse_count / len(foci)) * len(self.foci)
+                elif on_pulse_mismatch is OnPulseMismatchAction.ROUNDUP:
+                    solution_sequence.pulse_count = math.floor(self.sequence.pulse_count / len(foci)) * len(self.foci)
+                logging.warning(
+                    f"Pulse Count {self.sequence.pulse_count} is not a multiple of the number of foci {len(foci)}."
+                    f"Rounding to {solution_sequence.pulse_count}."
+                )
```

Comment:
done

Reply:
The separate function looks good, thanks. However the unit test needs to be improved, because currently it just copies the code that it is testing. For example in the unit test you have

```
if on_pulse_mismatch is OnPulseMismatchAction.ROUND:
    assert example_protocol.sequence.pulse_count == expected_pulse_count
    expected_pulse_count = round(example_protocol.sequence.pulse_count / len(foci)) * len(foci)
```

while in the code you have

```
if on_pulse_mismatch is OnPulseMismatchAction.ROUND:
    solution_sequence.pulse_count = round(self.sequence.pulse_count / len(self.foci)) * len(foci)
```

Of course the test will pass if you copy the same logic, but we still don't know that it behaves correctly.
We want to test the _behavior_ of the code, not just mirror the logic.
So here is how the unit test should work: Make a Protocol that has a Pulse in it that has a mismatch. Then run the `fix_pulse_mismatch` function and check that the pulse count becomes the actual number that we expect. Do this for each choice of `OnPulseMismatchAction`

Reply:
Same deal here, my suggestion is either to improve the unit test `test_fix_pulse_mismatch` or to remove it so that we understand that this is untested code. I agree with removing it due to our current time constraints.

Reply:
tried to rationnalized this one by creating a wheel pattern and testing against this one

Reply:
It looks great, thank you!